File: fsharp/4/Program.fs
open System

[<EntryPoint>]
let main argv =
    // Read the grid from stdin or a file
    let grid = Seq.initInfinite (fun _ -> Console.ReadLine())
                |> Seq.takeWhile (fun line -> line <> null)
                |> Seq.map (fun line -> line.ToCharArray())
                |> Seq.toArray

    let countXMAS (grid: char[][]) : int =
        let numRows = grid.Length
        let numCols = grid.[0].Length
        let directions = [| (-1, 0); (-1, 1); (0, 1); (1, 1); (1, 0); (1, -1); (0, -1); (-1, -1) |]
        let wordToFind = [| 'X'; 'M'; 'A'; 'S' |]
        let length = wordToFind.Length

        let isValidPosition x y =
            x >= 0 && x < numRows && y >= 0 && y < numCols

        let mutable count = 0
        for x in 0 .. numRows - 1 do
            for y in 0 .. numCols - 1 do
                for (dx, dy) in directions do
                    let mutable matchFound = true
                    let mutable i = 0
                    let mutable xi = x
                    let mutable yi = y
                    while matchFound && i < length do
                        if isValidPosition xi yi && grid.[xi].[yi] = wordToFind.[i] then
                            xi <- xi + dx
                            yi <- yi + dy
                            i <- i + 1
                        else
                            matchFound <- false
                    if matchFound then
                        count <- count + 1
        count

    let countXMAS_X (grid: char[][]) : int =
        let numRows = grid.Length
        let numCols = grid.[0].Length
        let patterns = Set.ofList [ "MAS"; "SAM" ]
        let length = 3 // Length of "MAS" or "SAM"

        let isValidPosition x y =
            x >= 0 && x < numRows && y >= 0 && y < numCols

        let mutable count = 0
        for x in 0 .. numRows - 3 do
            for y in 0 .. numCols - 3 do
                // Positions for the diagonals
                let x0, y0 = x, y
                let x1, y1 = x + 1, y + 1
                let x2, y2 = x + 2, y + 2

                let x3, y3 = x, y + 2
                let x4, y4 = x + 2, y

                if isValidPosition x0 y0 && isValidPosition x1 y1 && isValidPosition x2 y2 &&
                   isValidPosition x3 y3 && isValidPosition x4 y4 then

                    let diag1 = [| grid.[x0].[y0]; grid.[x1].[y1]; grid.[x2].[y2] |]
                    let diag2 = [| grid.[x3].[y3]; grid.[x1].[y1]; grid.[x4].[y4] |]

                    let word1 = System.String.Concat(diag1)
                    let word2 = System.String.Concat(diag2)
                    let word1Rev = System.String.Concat(Array.rev diag1)
                    let word2Rev = System.String.Concat(Array.rev diag2)

                    if (patterns.Contains word1 || patterns.Contains word1Rev) &&
                       (patterns.Contains word2 || patterns.Contains word2Rev) &&
                       diag1.[1] = 'A' && diag2.[1] = 'A' then
                        count <- count + 1
        count

    let countPart1 = countXMAS grid
    let countPart2 = countXMAS_X grid

    printfn "Part One: %d" countPart1
    printfn "Part Two: %d" countPart2

    0 // return an integer exit code

File: csharp/4/Program.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    static void Main(string[] args)
    {
        // Read the grid from standard input
        var grid = ReadGrid();

        // Part One: Count occurrences of "XMAS"
        int partOneCount = CountXMAS(grid);
        Console.WriteLine($"Part One: {partOneCount}");

        // Part Two: Count occurrences of "X-MAS" patterns
        int partTwoCount = CountXMAS_X(grid);
        Console.WriteLine($"Part Two: {partTwoCount}");
    }

    // Function to read the grid from input
    static char[][] ReadGrid()
    {
        var lines = new List<string>();
        string line;
        while ((line = Console.ReadLine()) != null && line != "")
        {
            lines.Add(line);
        }
        return lines.Select(l => l.ToCharArray()).ToArray();
    }

    // Part One: Count all occurrences of "XMAS" in any direction
    static int CountXMAS(char[][] grid)
    {
        int numRows = grid.Length;
        int numCols = grid[0].Length;
        string word = "XMAS";
        var directions = new (int dx, int dy)[]
        {
            (-1, 0),    // North
            (-1, 1),    // Northeast
            (0, 1),     // East
            (1, 1),     // Southeast
            (1, 0),     // South
            (1, -1),    // Southwest
            (0, -1),    // West
            (-1, -1)    // Northwest
        };

        var count = (
            from x in Enumerable.Range(0, numRows)
            from y in Enumerable.Range(0, numCols)
            from dir in directions
            where IsMatch(grid, x, y, dir.dx, dir.dy, word)
            select 1
        ).Sum();

        return count;
    }

    // Helper function to check if "XMAS" matches starting at (x, y) in direction (dx, dy)
    static bool IsMatch(char[][] grid, int x, int y, int dx, int dy, string word)
    {
        int numRows = grid.Length;
        int numCols = grid[0].Length;

        for (int i = 0; i < word.Length; i++)
        {
            int xi = x + i * dx;
            int yi = y + i * dy;

            if (xi < 0 || xi >= numRows || yi < 0 || yi >= numCols)
                return false;

            if (grid[xi][yi] != word[i])
                return false;
        }
        return true;
    }

    // Part Two: Count all "X-MAS" patterns
    static int CountXMAS_X(char[][] grid)
    {
        int numRows = grid.Length;
        int numCols = grid[0].Length;
        var patterns = new HashSet<string> { "MAS", "SAM" };

        var count = (
            from x in Enumerable.Range(0, numRows - 2)
            from y in Enumerable.Range(0, numCols - 2)
            let diag1 = new[]
            {
                grid[x][y],
                grid[x+1][y+1],
                grid[x+2][y+2]
            }
            let diag2 = new[]
            {
                grid[x][y+2],
                grid[x+1][y+1],
                grid[x+2][y]
            }
            let word1 = new string(diag1)
            let word2 = new string(diag2)
            let word1Rev = new string(diag1.Reverse().ToArray())
            let word2Rev = new string(diag2.Reverse().ToArray())
            where (patterns.Contains(word1) || patterns.Contains(word1Rev)) &&
                  (patterns.Contains(word2) || patterns.Contains(word2Rev)) &&
                  diag1[1] == 'A' && diag2[1] == 'A'
            select 1
        ).Sum();

        return count;
    }
}
